### ISSUE TITLE
Vast Glossy Tapir - invoke() early return eth

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-10-perennial-judging#37]
Vast Glossy Tapir

medium

# invoke() early return eth

## Summary
In `MultiInvoker.invoke()`, the return of the remaining `eth` is incorrectly placed inside the loop, resulting in the return of the remaining `eth` after the first `action`.
This results in the second and subsequent `actions` that need to use `eth`, there will be no `eth` available, and `invoke()` will `revert`

## Vulnerability Detail
in `MultiInvoker.invoke()`
If there is any `eth` left after `invocations` is executed, we should return the `eth` to the user to avoid `eth` remaining in the contract.
The code is as follows
```solidity
    function invoke(Invocation[] calldata invocations) external payable {
        for(uint i = 0; i < invocations.length; ++i) {
            Invocation memory invocation = invocations[i];
.......
            } else if (invocation.action == PerennialAction.COMMIT_PRICE) {
                (address oracleProviderFactory, uint256 value, bytes32[] memory ids, uint256 version, bytes memory data, bool revertOnFailure) =
                    abi.decode(invocation.args, (address, uint256, bytes32[], uint256, bytes, bool));

                _commitPrice(oracleProviderFactory, value, ids, version, data, revertOnFailure);
            } else if (invocation.action == PerennialAction.LIQUIDATE) {
                (IMarket market, address account, bool revertOnFailure) = abi.decode(invocation.args, (IMarket, address, bool));

                _liquidate(market, account, revertOnFailure);
            } else if (invocation.action == PerennialAction.APPROVE) {
                (address target) = abi.decode(invocation.args, (address));

                _approve(target);
            }

            // Eth must not remain in this contract at rest
@>          payable(msg.sender).transfer(address(this).balance);
        }
    }
```

As we can see from the code above, the current implementation of returning `eth` is placed inside the loop. 
This way the first `action` is executed and then the `eth` is returned.
If the `action` that requires `eth`, such as `PerennialAction.COMMIT_PRICE`, is the second `action`, by the time it's ready to be executed, but all the `eth` have already been returned, which will cause this `action` to fail to execute.

## Impact

Early return of `eth` will result in `actions` requiring `eth`  after the first `action` not being executed.

## Code Snippet

https://github.com/sherlock-audit/2023-10-perennial/blob/main/perennial-v2/packages/perennial-extensions/contracts/MultiInvoker.sol#L166

## Tool used

Manual Review

## Recommendation

```diff
    function invoke(Invocation[] calldata invocations) external payable {
        for(uint i = 0; i < invocations.length; ++i) {
            Invocation memory invocation = invocations[i];
...
            } else if (invocation.action == PerennialAction.APPROVE) {
                (address target) = abi.decode(invocation.args, (address));

                _approve(target);
            }

            // Eth must not remain in this contract at rest
-           payable(msg.sender).transfer(address(this).balance);
        }
+       payable(msg.sender).transfer(address(this).balance);
    }
```

**Comment by bin2chen66** [comment]
@cvetanovv Hi, please look at the message above from kbrizzle, does this count or not? thanks

**Comment by cvetanovv** [comment]
> @cvetanovv Hi, please look at the message above from kbrizzle, does this count or not? thanks

[bin2chen66](https://github.com/bin2chen66) There was a discussion whether the report should be Low or Medium. I agree with panprog comment and that is why I have not reopened the report despite the sponsor's wishes. Sherlock don't have a Low severity label and that's why there is confusion. However if you wish you can escalate.

**Comment by bin2chen66** [comment]
Escalate

Hi, cvetanovv, hope you can reconsider this issues,thanks

The current version and the `2023-07-perennial` version has a major difference in relation to the issue raised here: https://github.com/sherlock-audit/2023-07-perennial-judging/issues/23

The biggest difference is that users may not feel the `revert` in this version at all, because the new version has added `revertOnFailure`, which is not configured in `2023-07-perennial`. 
In the `2023-07-perennial` version, if the user submits multiple, it will `revert`, and the user realizes this problem and resubmits individually.

But in the current version, if `revertOnFailure=false` is set (the user's idea is: if a market is preemptively submitted, skip it and continue to submit other markets), because of the problem described, only the first one can always succeed. Importantly, users may find it difficult to realize that this bug is causing it, thinking that it is a normal phenomenon that has been preemptively submitted, resulting in the loss of keeper rewards in other markets.

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> Hi, cvetanovv, hope you can reconsider this issues,thanks
> 
> The current version and the `2023-07-perennial` version has a major difference in relation to the issue raised here: https://github.com/sherlock-audit/2023-07-perennial-judging/issues/23
> 
> The biggest difference is that users may not feel the `revert` in this version at all, because the new version has added `revertOnFailure`, which is not configured in `2023-07-perennial`. 
> In the `2023-07-perennial` version, if the user submits multiple, it will `revert`, and the user realizes this problem and resubmits individually.
> 
> But in the current version, if `revertOnFailure=false` is set (the user's idea is: if a market is preemptively submitted, skip it and continue to submit other markets), because of the problem described, only the first one can always succeed. Importantly, users may find it difficult to realize that this bug is causing it, thinking that it is a normal phenomenon that has been preemptively submitted, resulting in the loss of keeper rewards in other markets.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by nevillehuang** [comment]
> Escalate
> 
> Hi, cvetanovv, hope you can reconsider this issues,thanks
> 
> The current version and the `2023-07-perennial` version has a major difference in relation to the issue raised here: [sherlock-audit/2023-07-perennial-judging#23](https://github.com/sherlock-audit/2023-07-perennial-judging/issues/23)
> 
> The biggest difference is that users may not feel the `revert` in this version at all, because the new version has added `revertOnFailure`, which is not configured in `2023-07-perennial`. In the `2023-07-perennial` version, if the user submits multiple, it will `revert`, and the user realizes this problem and resubmits individually.
> 
> But in the current version, if `revertOnFailure=false` is set (the user's idea is: if a market is preemptively submitted, skip it and continue to submit other markets), because of the problem described, only the first one can always succeed. Importantly, users may find it difficult to realize that this bug is causing it, thinking that it is a normal phenomenon that has been preemptively submitted, resulting in the loss of keeper rewards in other markets.

@bin2chen66 it is important to note that in your original submission, the above mentioned impact was not mentioned. This made @panprog comments compelling, and accompanied by this original [comment in the previous audit](https://github.com/sherlock-audit/2023-07-perennial-judging/issues/23#issuecomment-1701164881), I think this could still be maintained as low severity.

**Comment by bin2chen66** [comment]
@nevillehuang 
Hi, nevillehuang, when I submit a question, I don’t consider the old questions. In fact, I didn’t even know about ‘2023-07-perennial-judging#23’ at the time. 
But the problem I described does exist, and I think it’s a problem, so I submit it.
It’s not possible for me to specifically look at all old questions and note the differences.

**Comment by nevillehuang** [comment]
Agreed, I’m just noting that the original submission seems to not qualify for a medium severity finding under sherlock guidelines if you see it as a DoS given the commit price action can be repeated. 

Could you shed more light on how keepers can possibly lose rewards?

**Comment by bin2chen66** [comment]
This problem is not DOS.

Assuming I am a keeper, I need to monitor the price of Oracle, and I have monitored the Oracle in 10 markets (note that this requires cost calculation).

I submitted the prices of 10 markets price to this contract, but only the first one was successful, and I got a reward, and I might think it's a normal phenomenon. Continue this in the scheduled task.
If there is no bug, I should get 10 rewards.
This is not right for users, this is a loss. Note that I have costs.

**Comment by panprog** [comment]
I agree that my comment and comparision to the previous contest doesn't fully apply to this issue in this contest due to ability to commit for multiple oracles and the revertOnFailure parameter. I'd say it's a borderline low/medium, because important protocol function might fail even if no loss of funds happen (keeper trying to commit prices for multiple markets will keep failing thus only 1 market will have fresh price commited). Given that the sponsor wanted to re-open it and considers it important to fix, I'm leaning towards medium, but I see valid arguments for low as well (the other keepers / the same keeper still commiting price for the other markets in the following blocks). Up to judge to decide.

**Comment by nevillehuang** [comment]
Hi @panprog is it true that no loss of funds will occur? If thats the case this could be kept as low severity given prices can be committed in subsequent blocks and keepers would receive the rewards subsequently. The only "loss" here is the loss of gas which is historically not accepted as valid findings in sherlock.

@bin2chen66 Does the delay in price updates affect the protocol in anyway? If yes I would like to hear the potential impact that can possibly reinstate this issue as Medium/High.

**Comment by bin2chen66** [comment]
@nevillehuang
Hi, the price is constantly delayed, like repeatedly submitting: 10->9->8->7... The last few markets may time out and become invalid prices. 
```solidity
    /// @dev After this amount of time has passed for a version without being committed, the version can be invalidated.
    uint256 public immutable timeout;
```
Invalid prices will cause all positions at this timestamp in this market to become invalid.

However, it's hard to say whether this will happen, because it involves the number of keeper bots and how the bot implements this task, and when the keeper discovers this bug and adjusts the code.

**Comment by panprog** [comment]
@nevillehuang 
1. If Keeper1 tries to commit oracle1 and oracle2, only oracle1 is commited, oracle2 is not commited (Keeper1 doesn't receive reward for it). In the following transaction Keeper2 commits oracle2 and gets reward for it, meaning Keeper1 has lost his reward for Oracle2. Not sure if it qualifies as a loss of funds for Keeper1 (as the funds are not lost, just given to a different keeper)?
2. What @bin2chen66 said: if there are a lot of markets to commit price to, but each transaction only the 1st of them will get commited, transactions commiting the last markets might hit timeout, causing the oracle version to be invalid, which can cause all kinds of problems for the protocol (even though invalid oracle versions are normal protocol operations, it still causes problems for users including loss of funds and must be a very rare event. If it's frequent, that's a problem). However, I don't think this is possible in real deployment and might also qualify as wrong timeout setting (admin error): if timeout is set reasonably, this will require like 50+ oracle updates in a single call, which is hardly possible due to gas constraints. If we take some realistic number, like 10 oracle updates in a single call - timeout has to be about 10-20 seconds to increase probability of invalid version. Besides, it's not like there is just 1 keeper, the other keepers should also be able to commit the other markets in the same block, further decreasing the chances of invalid oracle version happening.

I'm leaning towards medium not based on loss of funds, but based on incorrect behavior of rather important (core) protocol function, even if no real loss of funds happen.

**Comment by nevillehuang** [comment]
Hi @panprog based on your above comments, I am definitely leaning towards medium severity just based off point 1 alone as I indeed see it as loss of rewards for keeper 1, unless it is previously make known that keepers are admin controlled entities.

**Comment by panprog** [comment]
Anyone can be liquidator, so it's definitely not admin controlled entities.

**Comment by cvetanovv** [comment]
[nevillehuang](https://github.com/nevillehuang) After reading all the comments and the sponsor desire to reopen the report I agree with bin2chen and panprog reasoning and this should be a valid Medium

**Comment by nevillehuang** [comment]
Agree, suggest to keep this issue as medium, unless @Czar102 has an alternative opinion. But would encourage the watson to explore the full impact of the finding in future submissions, since imo it is not the judges responsibility to do so.

**Comment by Czar102** [comment]
Hi, I'm sorry for the delay on my comment on this issue.

About @panprog's points:
1. I'm not sure about the role keepers have in the system but if it's like solvers in intent-based exchanges (that is, they are paid for the work done), then I don't see any loss of funds happening there. In other words, another party taking the incentive to execute an action does not constitute a loss of funds, it's just an offchain actor working suboptimally.
2. When someone sets `revertOnFailure` parameter to `false`, they should account for the fact that the action may revert. I'm not sure if it's fair to assume that a revert means that the oracle feed was already submitted, can I have some more details about this assumption @cvetanovv @nevillehuang @bin2chen66?

Additionally, the initial sponsor's claim seems severe, can you confirm it or add more details why wasn't it further discussed to judge this issue as a Medium @cvetanovv @nevillehuang @bin2chen66? From my understanding, the sponsor states that this impacted functionality is critical for some use cases and cannot be worked around.
> being able to commit 2+ prices atomically is required by the protocol (for example settling multiple markets for a vault operation)

**Comment by nevillehuang** [comment]
@Czar102 

1. If there is an error in the implementation of code logic that results in an existing keeper getting the fees intended for the original keeper, wouldn't that constitute as a loss of rewards due to broken functionality?

2. Not sure where the assumption came from that reverting means oracle feed is already submitted. From my understanding reverting means oracle feed has not been submitted and keeper needs to re-commit prices.

3. Agree, @kbrizzle could you shed more light on what you meant when you mention it is necessary to atomically commit 2+ prices to settle multiple markets? Specifically, which areas of the code would be affected, because if core functionality is affected that leads to potential loss of funds (depending on impact) the severity could be high.

However, it is still important to note that the watson did not explore the full impact of the finding and simply stated a potential revert, so even if there is additional impact discovered, I think we should still maintain medium severity.

**Comment by panprog** [comment]
@Czar102 
1. Yes, keepers are paid for the job they do (transaction fees+some bonus). The issue causes the keeper who correctly submits more than 1 job in 1 transacton either revert (thus not receiving any incentive) or receive incentive only for the 1st job (even though the other jobs are submitted correctly). So off-chain actor does everything correctly, but doesn't receive full reward for its actions.
2. `revertOnFailure` just makes it possible that the 1st job is executed and rewarded while all the other jobs submitted silently fail (without it - transaction will simply revert). Not sure which one is more severe - I think both options are similar in that the core functionality fails, just in different ways.
3. As I understand, they mean that when any vault actions happens, protocol wants to ensure that all the underlying price market oracles are commited to the latest available price. For example, if vault has ETH and BTC markets, then before any vault action protocol wants to have latest oracle version commited both for ETH and BTC, so it wants 3 actions in 1 transaction: ETH commit, BTC commit, vault action. The issue described here can lead to: ETH commited, BTC commit silent failure, vault action (with the latest ETH market price, but not BTC market price). However, this is not something that must happen (just what's more beneficial for the transaction flow) and at most can lead to vault action revert if the user's previous action is not yet settled. So it's surely not high severity and I'm not sure if it qualifies for medium (no real loss of funds, at most transaction revert).

**Comment by bin2chen66** [comment]
Hi , @Czar102

Here is my personal understanding:

Firstly, the worst-case scenario of this issue is that it results in an invalid price, which subsequently invalidates the following positions. However, as I mentioned earlier, this situation is likely to occur with a very small probability, so it does not reach the level of High severity.

I personally think that Medium severity is reasonable, based on the following two points:
- Misallocation of rewards (The user did nothing wrong, but due to the protocol bug, they did not receive the deserved rewards. I think this is a kind of loss for the user.)
- Loss of functionality (According to the sponsor's comments, this function is not an optional function. Due to this bug, this function cannot work properly. It is more cost-effective to go directly to the original oracle.sol and submit one by one.)

thanks.






Note: Assuming you submit prices for 10 markets, I think the one import function of `revertOnFailure` is that if someone submits to a market before you, that market will revert, but it will be ignored, and you can still receive rewards for the other 9 markets. However, the problem caused by this bug is that even if you are always the first to submit, you can only get the first one, and the other 9 cannot be obtained in this transaction (of course, you can submit the transaction again, but whether it will be you get rewards is hard to say).

**Comment by Czar102** [comment]
> If there is an error in the implementation of code logic that results in an existing keeper getting the fees intended for the original keeper, wouldn't that constitute as a loss of rewards due to broken functionality?

No, in the same way as solvers being unable to include a swap transactions aren't losing funds because they would get a reward if the bug didn't prevent them from inclusion.

"Misallocation of rewards" here is a misallocation of means, not ends. Hence I don't consider it an issue at all. The script, frontend, etc. should detect the revert of the transaction by simulating it beforehand, which won't cause any loss of funds. Users are ignoring the reverts at their own risk when using `!revertOnFailure`.

I see @panprog's 3rd point. If that's the case, it is a valid low since it causes a partial loss of the functionality (it is still possible to do the same, just in more steps).

I would still like to get sponsor's confirmation on this, whether committing 2+ prices in a single call is not strictly needed @kbrizzle. If the answer is going to agree with @panprog's, I intend to leave this issue as is. Also, the full impact hasn't been explored in the initial submission.

**Comment by arjun-io** [comment]
While not strictly _required_ per se, the UX degrades drastically (almost to the point of being unusable) if this bug were to persist. Essentially, if markets have not been updated in a while when a user goes to deposit to a vault, they would need to send N separate transactions (where N is the number of markets supported by the vault) within 60s in order to get all markets up to date before the vault deposit can process.

A lot of the discussion has been on the keeper side, where I agree with @panprog that keeper submissions don't need to be atomic.

**Comment by Czar102** [comment]
Result:
Low
Unique

Based on the comments above, I can certainly say that there is no loss of funds. The UX is very bad, but the contracts remain (partially) functional.

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [bin2chen66](https://github.com/sherlock-audit/2023-10-perennial-judging/issues/37/#issuecomment-1823694477): rejected

**Comment by arjun-io** [comment]
Fixed: https://github.com/equilibria-xyz/perennial-v2/pull/157

**Comment by panprog** [comment]
Mitigation Review:

Fixed
